pkg/policy/mapstate.go

```diff
@@ -100,3 +101,74 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
 		keys[keyToAdd] = MapStateEntry{}
 	}
 }
+
+// MapChanges collects updates to the endpoint policy on the
+// granularity of individual mapstate key-value pairs for both adds
+// and deletes. 'mutex' must be held for any access.
+type MapChanges struct {
+	mutex   lock.Mutex
+	adds    MapState
+	deletes MapState
+}
+
+// AccumulateMapChanges accumulates the given changes to the
+// MapChanges, updating both maps for each add and delete, as
+// applicable.
+func (mc *MapChanges) AccumulateMapChanges(adds, deletes []identity.NumericIdentity,
+	port uint16, proto uint8, direction trafficdirection.TrafficDirection) {
+	key := Key{
+		// The actual identity is set in the loops below
+		Identity: 0,
+		// NOTE: Port is in host byte-order!
+		DestPort:         port,
+		Nexthdr:          proto,
+		TrafficDirection: direction.Uint8(),
+	}
+	value := MapStateEntry{
+		ProxyPort: 0, // Will be updated by the caller when applicable
+	}
+
+	log.Debugf("MapChanges: AccumulateMapChanges(adds: %v, deletes: %v, port: %d, proto: %d, direction: %d)",
+		adds, deletes, port, proto, direction.Uint8())
+
+	mc.mutex.Lock()
+	if len(adds) > 0 {
```

Comment:
The sets 'adds' and 'deletes' are disjoint so that no identity is present in both. I added a commit documenting this and verified that this is the case on all call sites.